cmd/deploy.go

```diff
@@ -225,33 +226,27 @@ func createHostsFile(containers []types.Container, bridgeName string) error {
 }
 
 // hostEntries builds an /etc/hosts compliant text blob (as []byte]) for containers ipv4/6 address<->name pairs
-func hostsEntries(containers []types.Container, bridgeName string) []byte {
+func hostsEntries(containers []types.GenericContainer, bridgeName string) []byte {
 	buff := bytes.Buffer{}
 	for _, cont := range containers {
 		if len(cont.Names) == 0 {
 			continue
 		}
-		if cont.NetworkSettings != nil {
-			if br, ok := cont.NetworkSettings.Networks[bridgeName]; ok {
-				if br.IPAddress != "" {
-					buff.WriteString(br.IPAddress)
-					buff.WriteString("\t")
-					buff.WriteString(strings.TrimLeft(cont.Names[0], "/"))
-					buff.WriteString("\n")
-				}
-				if br.GlobalIPv6Address != "" {
-					buff.WriteString(br.GlobalIPv6Address)
-					buff.WriteString("\t")
-					buff.WriteString(strings.TrimLeft(cont.Names[0], "/"))
-					buff.WriteString("\n")
-				}
-			}
+		if cont.NetworkSettings.Set {
+			buff.WriteString(cont.NetworkSettings.IPv4addr)
+			buff.WriteString("\t")
+			buff.WriteString(strings.TrimLeft(cont.Names[0], "/"))
+			buff.WriteString("\n")
+			buff.WriteString(cont.NetworkSettings.IPv6addr)
+			buff.WriteString("\t")
+			buff.WriteString(strings.TrimLeft(cont.Names[0], "/"))
+			buff.WriteString("\n")
```

Comment:
there might be a situation where IPv6 addresses are not set while IPv4 addresses are (and vice versa)

for example, if you use the following mgmt config:

```yaml
name: ipv4only
mgmt:
  ipv4_subnet: 172.30.30.0/24
```

then your nodes will be addressed with IPv4 addresses only from the specified range. At the same time, the current function will add hosts entires for empty ipv6 addresses, it seems

Reply:
I don't fully understand this. Is this determined inside the`initMgmtNetwork` ?

```
	if c.Config.Mgmt.IPv4Subnet == "" && c.Config.Mgmt.IPv6Subnet == "" {
		if c.Config.Mgmt.IPv4Subnet == "" {
			c.Config.Mgmt.IPv4Subnet = dockerNetIPv4Addr
		}
		if c.Config.Mgmt.IPv6Subnet == "" {
			c.Config.Mgmt.IPv6Subnet = dockerNetIPv6Addr
		}
	}
```

Reply:
but that's exactly how this function works now. TBH, the extra if can be dropped as it's duplicate:
```
	if c.Config.Mgmt.IPv4Subnet == "" && c.Config.Mgmt.IPv6Subnet == "" {
		c.Config.Mgmt.IPv4Subnet = dockerNetIPv4Addr
		c.Config.Mgmt.IPv6Subnet = dockerNetIPv6Addr
	} 
```
That would set the default values only when both v4 and v6 subnets are missing. In all other cases, they will be populated with user-provided values.

Is this a purely incorrect reporting bug?

Reply:
@networkop the culprit was in moving `initMgmtNetwork` from `ParseTopology` to `NewContainerlab` here - https://github.com/srl-labs/containerlab/pull/416/files#diff-831ecb8143ceda28e72434a64e3b7f88d15f9492071603608fd0180ae3105e67R88-R92

it led to reading the `mgmt` config element before it was parsed from the clab file


Let's move it back to `ParseTopology` as it [was](https://github.com/srl-labs/containerlab/blob/169c8bdb33d78a6bb5006d568dbc396561af8eb9/clab/config.go#L229) before?

Reply:
interesting. the reason why I moved it out was what we're discussing above - interaction with runtime. Let's assume I put it back into ParseTopology, I would then have to call `runtime.SetMgmtNet` with `c.Config.Mgmt`.  I can't call it inside `WithRuntime` because topology hasn't been parsed yet, so the only place to do that is inside each of the `./cmd/<action>.go` files, similar to this:

```
if err = c.ParseTopology(); err != nil {
	return err
}
c.Runtime.SetMgmtNet(c.Config.Mgmt)
```

This didn't look right to me at the time, as I was trying to minimise the impact of the change. Does this look ok to you?

Reply:
alternatively we might put the logic that sets the management network defaults inside the `WithTopoFile` func, it reads the clab yaml file and then sets the necessary defaults

```diff
func WithTopoFile(file string) ClabOption {
	return func(c *CLab) {
		if file == "" {
			return
		}
		if err := c.GetTopology(file); err != nil {
			log.Fatalf("failed to read topology file: %v", err)
		}

+		err := c.initMgmtNetwork()
+		if err != nil {
+			log.Fatalf("initMgmtNetwork: %s", err)
+		}
	}
}
```

It is not "clean" logic-wise, but seems to do the job

Reply:
cool. check the latest PR

Reply:
coming back to the original issue using the latest commit:

the current code for writing host entries doesn't account for the fact that ipv6/4 address might be empty, this leads to the following lines created when only ipv4_subnet is specified:

```
172.20.30.2	clab-srl01-srl
	clab-srl01-srl
```

the original code in maste checks first if the ipv4/6 address is not `""` and then writes the host entry, this I believe should be used here as well to omit the entries which do not have ipv4/6 address

Reply:
oh yeah, must have missed that.